nat/service_ics.go

```diff
@@ -0,0 +1,234 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package nat
+
+import (
+	"net"
+	"os/exec"
+	"strconv"
+	"strings"
+	"sync"
+
+	log "github.com/cihub/seelog"
+	"github.com/pkg/errors"
+)
+
+const (
+	enablePublicSharing  = "$config.EnableSharing(0)"
+	enablePrivateSharing = "$config.EnableSharing(1)"
+	disableSharing       = "$config.DisableSharing()"
+)
+
+type serviceICS struct {
+	mu                 sync.Mutex
+	ifaces             map[string]RuleForwarding // list in internal interfaces with enabled internet connection sharing
+	remoteAccessStatus string
+	powerShell         func(cmd string) ([]byte, error)
+}
+
+// Enable enables internet connection sharing for the public interface.
+func (nat *serviceICS) Enable() error {
+	status, err := nat.powerShell("Get-Service RemoteAccess | foreach { $_.StartType }")
+	if err != nil {
+		return errors.Wrap(err, "failed to get RemoteAccess service startup type")
+	}
+	nat.remoteAccessStatus = string(status)
+
+	if _, err := nat.powerShell("Set-Service -Name RemoteAccess -StartupType automatic"); err != nil {
+		return errors.Wrap(err, "failed to set RemoteAccess service startup type to automatic")
+	}
+	if _, err := nat.powerShell("Start-Service -Name RemoteAccess"); err != nil {
+		return errors.Wrap(err, "failed to start RemoteAccess service")
+	}
+
+	ifaceName, err := nat.getPublicInterfaceName()
+	if err != nil {
+		return errors.Wrap(err, "failed to get public interface name")
+	}
+
+	err = nat.applySharingConfig(enablePublicSharing, ifaceName)
+	return errors.Wrap(err, "failed to enable internet connection sharing")
+}
+
+// Add enables internet connection sharing for the local interface.
+func (nat *serviceICS) Add(rule RuleForwarding) error {
+	// TODO firewall rule configuration should be added here for new connections.
+	ifaceName, err := getInterfaceBySubnet(rule.SourceAddress)
+	if err != nil {
+		return errors.Wrap(err, "failed to find suitable interface")
+	}
+
+	err = nat.applySharingConfig(enablePrivateSharing, ifaceName)
+	if err != nil {
+		return errors.Wrap(err, "failed to enable internet connection sharing for internal interface")
+	}
+
+	nat.mu.Lock()
+	defer nat.mu.Unlock()
+	nat.ifaces[ifaceName] = rule
+
+	return nil
+}
+
+// Del disables internet connection sharing for the local interface.
+func (nat *serviceICS) Del(rule RuleForwarding) error {
+	// TODO firewall rule configuration should be added here for cleaning up unused connections.
+	ifaceName, err := getInterfaceBySubnet(rule.SourceAddress)
+	if err != nil {
+		return errors.Wrap(err, "failed to find suitable interface")
+	}
+
+	err = nat.applySharingConfig(disableSharing, ifaceName)
+	if err != nil {
+		return errors.Wrap(err, "failed to disable internet connection sharing for internal interface")
+	}
+
+	nat.mu.Lock()
```

Comment:
Should not locking be before actions with interfaces?

Reply:
This lock is just protecting map from concurrent access. 
If we want to have locks for the whole configuration process, it will require more changes.